### PR TITLE
Refactor redirects index view to CBV

### PR DESCRIPTION
**Pull request by rohitsrma**
In this PR I've converted redirects index view from FBV to CBV as a part of https://github.com/wagtail/wagtail/issues/8365.

Build on top of #11035 

image

### CODE REVIEW

**Comment by rohitsrma** [comment]
Hello @laymonage, could you please help me understand why I am getting these errors?

```
SystemCheckError: System check identified some issues:

ERRORS:
tests.FullFeaturedSnippet: (models.E025) The property 'revisions' clashes with a related field accessor.
tests.FullFeaturedSnippet: (models.E025) The property 'workflow_states' clashes with a related field accessor.
```

Locally everything is working fine!

**Comment by laymonage** [comment]
Thanks, @rohitsrma! Don't worry about it, it's a bug in the latest Django `main` branch – see #11746.

**Comment by rohitsrma** [comment]
Hello @laymonage, I was hoping to get your guidance on something.

In this PR, I've replaced the `list.html` with our tables UI framework. However, the `Export Redirects` section still relies on this file (list.html). I tried to use the framework there too, but it seems we'd need to make adjustments to the filters for them to work correctly. Could you advise on the best approach to proceed?

File: wagtail/contrib/redirects/views.py [hunk 1]
```diff
@@ -36,65 +36,91 @@
...
+        TypeColumn(
+            "get_is_permanent_display",
+            label=gettext_lazy("Type"),
+            width="15%",
+        ),
+    ]
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.redirects = Redirect.objects.prefetch_related("redirect_page", "site")
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:101]: Why is this here? Can this be in `get_queryset` or `get_base_queryset` instead?

File: wagtail/contrib/redirects/views.py [hunk 2]
```diff
@@ -36,65 +36,91 @@
...
+        self.query_string = request.GET.get("q", "")
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:102]: This is already available as `self.search_query` from the generic view

File: wagtail/contrib/redirects/views.py [hunk 3]
```diff
@@ -36,65 +36,91 @@
...
+
+    def get_queryset(self):
+        redirects = self.redirects
+        if self.is_searching:
+            if self.query_string:
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:107]: This won't be necessary if we use `self.search_query`
```suggestion
```

File: wagtail/contrib/redirects/views.py [hunk 4]
```diff
@@ -36,65 +36,91 @@
...
+                )
+
+        if self.get_ordering() == "old_path":
+            redirects = redirects.order_by(self.default_ordering)
+
+        return redirects
+
+    def get_context_data(self):
+        context_data = super().get_context_data()
+        context_data["ordering"] = self.default_ordering
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:121]: Is this used in the template?

```suggestion
```

File: wagtail/contrib/redirects/views.py [hunk 5]
```diff
@@ -36,65 +36,91 @@
...
+        context_data["redirects"] = self.get_queryset()
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:122]: Is it possible to use `context_object_name` on the view class itself?

File: wagtail/contrib/redirects/views.py [hunk 6]
```diff
@@ -36,65 +36,91 @@
...
+        self.query_string = request.GET.get("q", "")
+
+    def get_queryset(self):
+        redirects = self.redirects
+        if self.is_searching:
+            if self.query_string:
+                redirects = redirects.filter(
+                    Q(old_path__icontains=self.query_string)
+                    | Q(redirect_page__url_path__icontains=self.query_string)
+                    | Q(redirect_link__icontains=self.query_string)
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:111]: This can probably be replaced by setting `search_fields = ["old_path", "redirect_page__url_path", "redirect_link"]` on the view class (not sure if it supports related fields though)

File: wagtail/contrib/redirects/views.py [hunk 7]
```diff
@@ -36,65 +36,91 @@
...
 
-@permission_checker.require_any("add", "change", "delete")
-@vary_on_headers("X-Requested-With")
-def index(request):
-    query_string = request.GET.get("q", "")
-    ordering = request.GET.get("ordering", "old_path")
+    def get_cell_context_data(self, instance, parent_context):
+        context = super().get_cell_context_data(instance, parent_context)
+        context["redirects"] = instance
+        return context
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:46]: The `instance` is already available as the `instance` variable in the context from the `Column.get_cell_context_data`

```suggestion
```

**Review by laymonage — COMMENTED** [review]
Thanks @rohitsrma, this is a nice start.

The "exports" view is a report view and it doesn't support the new UI yet, so just leave it as it is. Don't worry about the `list.html`, we can keep it in the codebase for now. We can make it so that the index view itself use the tables framework, but let the reports view use the `list.html` template.

**Comment by rohitsrma** [comment]
Thanks @laymonage, I've made the changes.

image

I'm unsure why the checks are failing. Are they related to the changes I made?

File: wagtail/contrib/redirects/templates/wagtailredirects/type.html [hunk 8]
```diff
@@ -0,0 +1,2 @@
+{% load i18n wagtailadmin_tags %}
+<td class="type">{% status value classname="primary"|yesno:"w-status--primary," %}</td>
```
[the remark this answers is not in the document]
Comment (rohitsrma) [reply wagtail/contrib/redirects/templates/wagtailredirects/type.html]: Thanks, I've removed this file!

File: wagtail/contrib/redirects/views.py (hunk 1 above)
Comment (rohitsrma) [reply wagtail/contrib/redirects/views.py]: I tried this previously, but it was giving an error related to `request`. However, now I've completely removed it.

**Comment by gasman** [comment]
@rohitsrma The failing test is due to https://github.com/wagtail/wagtail/pull/11815/files#r1556095493 . I'm on it...

**Comment by rohitsrma** [comment]
@gasman, thanks for letting me know!

File: wagtail/contrib/redirects/views.py (hunk 4 above)
Comment (rohitsrma) [reply wagtail/contrib/redirects/views.py]: Sorry, it's a mistake!

File: wagtail/contrib/redirects/templates/wagtailredirects/index.html [hunk 9]
```diff
@@ -4,12 +4,10 @@
...
-
 {% block content %}
     {% trans "Redirects" as redirects_str %}
-    {% url 'wagtailredirects:index' as search_results_url %}
-    {% if user_can_add %}
+    {% url 'wagtailredirects:index_results' as search_results_url %}
+    {% if can_add %}
         {% url "wagtailredirects:add" as add_link %}
         {% trans "Add redirect" as add_str %}
         {% url "wagtailredirects:start_import" as import_link %}
```
[the remark this answers is not in the document]
Comment (rohitsrma) [reply wagtail/contrib/redirects/templates/wagtailredirects/index.html]: I've used this with little tweaks. Would love to hear your thoughts on the changes!

File: wagtail/contrib/redirects/templates/wagtailredirects/results.html [hunk 10]
```diff
@@ -1,23 +1,19 @@
...
         {% url 'wagtailredirects:add' as wagtailredirects_add_redirect_url %}
-        <p>{% blocktrans trimmed %}No redirects have been created. Why not <a href="{{ wagtailredirects_add_redirect_url }}">add one</a>?{% endblocktrans %}</p>
+        <p>
+            {% blocktrans trimmed %}
+                No redirects have been created. Why not <a href="{{ wagtailredirects_add_redirect_url }}">add one</a>?
+            {% endblocktrans %}
+        </p>
     {% endif %}
-{% endif %}
+{% endblock no_results_message %}
```
Comment (laymonage) [inline wagtail/contrib/redirects/templates/wagtailredirects/results.html:19]: We generally don't use the block name in `endblock` tags. Also, please make sure that the files end with a new line.

```suggestion
{% endblock %}

```

File: wagtail/contrib/redirects/templates/wagtailredirects/results.html [hunk 11]
```diff
@@ -1,23 +1,19 @@
...
+{% block no_results_message %}
     {% if query_string %}
-        <p role="alert">{% blocktrans trimmed %}Sorry, no redirects match "<em>{{ query_string }}</em>"{% endblocktrans %}</p>
+        <p role="alert">
+            {% blocktrans trimmed %}
+                Sorry, no redirects match "<em>{{ query_string }}</em>"
+            {% endblocktrans %}
+        </p>
     {% else %}
         {% url 'wagtailredirects:add' as wagtailredirects_add_redirect_url %}
```
Comment (laymonage) [inline wagtail/contrib/redirects/templates/wagtailredirects/results.html:12]:
```suggestion
```

File: wagtail/contrib/redirects/templates/wagtailredirects/results.html [hunk 12]
```diff
@@ -1,23 +1,19 @@
...
-        <p>{% blocktrans trimmed %}No redirects have been created. Why not <a href="{{ wagtailredirects_add_redirect_url }}">add one</a>?{% endblocktrans %}</p>
+        <p>
+            {% blocktrans trimmed %}
```
Comment (laymonage) [inline wagtail/contrib/redirects/templates/wagtailredirects/results.html:14]:
```suggestion
            {% blocktrans trimmed with wagtailredirects_add_redirect_url=add_url %}
```

File: wagtail/contrib/redirects/templates/wagtailredirects/redirect_page.html [hunk 13]
```diff
@@ -0,0 +1,8 @@
+{% load i18n wagtailadmin_tags %}
+<td>
+    {% if redirects.redirect_page %}
+        <a href="{% url 'wagtailadmin_pages:edit' redirects.redirect_page.id %}" class="nolink">{{ redirects.redirect_page }}</a>
+    {% else %}
+        {{ redirects.redirect_link }}
```
Comment (laymonage) [inline wagtail/contrib/redirects/templates/wagtailredirects/redirect_page.html:6]:
```suggestion
    {% if instance.redirect_page %}
        <a href="{% url 'wagtailadmin_pages:edit' instance.redirect_page_id %}" class="nolink">{{ instance.redirect_page }}</a>
    {% else %}
        {{ instance.redirect_link }}

```

File: wagtail/contrib/redirects/views.py (hunk 7 above)
Comment (laymonage) [reply wagtail/contrib/redirects/views.py:47]: It looks like this wasn't committed yet?...

File: wagtail/contrib/redirects/views.py [hunk 14]
```diff
@@ -36,66 +37,88 @@
 )
 from wagtail.log_actions import log
 
-permission_checker = PermissionPolicyChecker(permission_policy)
 
+class RedirectPageColumn(Column):
+    cell_template_name = "wagtailredirects/redirect_page.html"
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:42]: Could we use the `_cell` suffix for the template name? Also, since the redirect target may not be a page, maybe name this `RedirectToColumn` or `RedirectTargetColumn`, e.g.

```suggestion
class RedirectTargetColumn(Column):
    cell_template_name = "wagtailredirects/redirect_target_cell.html"
```

File: wagtail/contrib/redirects/views.py [hunk 15]
```diff
@@ -36,66 +37,88 @@
...
+            Button(
+                _("Export redirects"),
+                url=reverse("wagtailredirects:report"),
+                icon_name="download",
+                priority=100,
+            )
         )
 
+        return buttons
+
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:121]: The optimisation was lost. Also, maybe we could use `select_related` instead of `prefetch_related`

```suggestion

def get_base_queryset(self):
    return super().get_base_queryset().select_related("redirect_page", "site")

```

File: wagtail/contrib/redirects/views.py [hunk 16]
```diff
@@ -1,23 +1,23 @@
...
-from django.utils.translation import ngettext
+from django.utils.translation import gettext_lazy, ngettext
 from django.views.decorators.http import require_http_methods
-from django.views.decorators.vary import vary_on_headers
 
 from wagtail.admin import messages
 from wagtail.admin.auth import PermissionPolicyChecker
 from wagtail.admin.forms.search import SearchForm
+from wagtail.admin.ui.tables import Column, StatusTagColumn, TitleColumn
+from wagtail.admin.views.generic import IndexView
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:18]: So we can use the name `IndexView` instead of `Index`

```suggestion
from wagtail.admin.views import generic
```

File: wagtail/contrib/redirects/views.py [hunk 17]
```diff
@@ -36,66 +37,88 @@
...
-            | Q(redirect_link__icontains=query_string)
-        )
+permission_checker = PermissionPolicyChecker(permission_policy)
 
-    # Ordering (A bit useless at the moment as only 'old_path' is allowed)
-    if ordering not in ["old_path"]:
-        ordering = "old_path"
 
-    redirects = redirects.order_by(ordering)
+class Index(IndexView):
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:53]:
```suggestion
class IndexView(generic.IndexView):
```

File: wagtail/contrib/redirects/views.py [hunk 18]
```diff
@@ -36,66 +37,88 @@
...
+    template_name = "wagtailredirects/index.html"
+    results_template_name = "wagtailredirects/results.html"
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:55]: Let's rename the template while we're here

```suggestion
    results_template_name = "wagtailredirects/index_results.html"
```

**Review by laymonage — COMMENTED** [review]
Thanks! This is almost there... 😄

**Comment by rohitsrma** [comment]
Thanks @laymonage, I've made the changes. Please let me know if any further updates are needed.

File: wagtail/contrib/redirects/views.py (hunk 15 above)
Comment (rohitsrma) [reply wagtail/contrib/redirects/views.py:119]: Sorry about that, it's fixed now. I don't know why I do these dumb things sometimes. 😅

File: wagtail/contrib/redirects/views.py (hunk 7 above)
Comment (rohitsrma) [reply wagtail/contrib/redirects/views.py]: Sorry, I forgot to make this change - it's fixed now.

File: wagtail/contrib/redirects/templates/wagtailredirects/list.html [hunk 19]
```diff
@@ -1,43 +1,43 @@
...
-                    {% else %}
-                        {{ redirect.link }}
-                    {% endif %}
-                </td>
-                <td class="type">{% status redirect.get_is_permanent_display classname=primary|yesno:"w-status--primary," %}</td>
-            </tr>
-        {% endfor %}
-    </tbody>
-</table>
+{% load i18n wagtailadmin_tags %}
```
[the remark this answers is not in the document]
Comment (rohitsrma) [reply wagtail/contrib/redirects/templates/wagtailredirects/list.html]: After messing up my branch, I've reverted it back. 🤦‍♂️😁

File: wagtail/contrib/redirects/views.py [hunk 20]
```diff
@@ -39,63 +40,83 @@
...
+            label=gettext_lazy("From"),
+            width="35%",
+            url_name="wagtailredirects:edit",
+            sort_key="old_path",
+        ),
+        Column(
+            "site",
+            label=gettext_lazy("Site"),
+            width="25%",
+            sort_key="site",
```
Comment (laymonage) [inline wagtail/contrib/redirects/views.py:78]: This would be more useful if we use the site's `site_name` field, so it doesn't order by the site's primary key.

```suggestion
            sort_key="site__site_name",
```

**Review by laymonage — COMMENTED** [review]
Only minor comments left, but I'll fix them as I rebase+merge. Thanks!

**Review by laymonage — APPROVED** [review]
